client/state/data-layer/wpcom/concierge/schedules/initial/from-api.js

```diff
@@ -0,0 +1,22 @@
+/** @format */
+
+/**
+ * Internal dependencies
+ */
+import makeJsonSchemaParser from 'lib/make-json-schema-parser';
+import responseSchema from './schema';
+import { transform as appointmentTransformer } from 'state/data-layer/wpcom/concierge/schedules/appointments/detail/from-api';
+
+export const convertToDate = timestampInSeconds => timestampInSeconds * 1000;
```

Comment:
I am having a hard time understanding the purpose of the `convertToDate` function - I both can't imagine why would we need to multiply seconds 1000 and what “date” in the name means.

Reply:
> Isn't is possible the schema to also be reused?

We could but I think it would make the code a bit more complex as we'd have to import the schema and then combine them.

The added complexity in that would lie in `next_appointment` - the value for this could be an `object` or it could be `null`. In the schema I can just have `"type": [ "object", "null" ],` but if I were to import it I'd have to either modify the imported schema on the fly, updating the type to look like the previous example, or conditionally set it as either `null` or the imported schema depending on if a next appointment is available.

> I am having a hard time understanding the purpose of the convertToDate function - I both can't imagine why would we need to multiply seconds 1000 and what “date” in the name means.

The multiplication is because JavaScript handles Unix timestamps in milliseconds rather than seconds. A timestamp of `1531386772` in JavaScript would be 18th January 1970, for example, so we have to convert timestamps received from the API to account for this. I haven't expressly document this conversion as it's normal JavaScript behaviour, but I'm happy to go back and add a JSDoc if you feel that'd be useful to others?

As for the naming - I've used the same method used elsewhere in the Concierge code so I've kept with the naming convention. I would imagine it relates to why the timestamp is being converted - so we can work with JavaScript date objects. Not at all married to the naming though. Perhaps renaming it `convertToMilliseconds` will clear up both of these concerns in one go?

Reply:
> it would make the code a bit more complex

What are the trade-offs of reusing + more complex code vs. simpler and duplicated? What should we take into account in order to make a decision?

This is not a loaded question, I honestly don't know which ones is the better solution, but I would like us to think a bit more about what's best.

Reply:
> As for the naming - I've used the same method used elsewhere in the Concierge code so I've kept with the naming convention. I would imagine it relates to why the timestamp is being converted - so we can work with JavaScript date objects. Not at all married to the naming though. Perhaps renaming it `convertToMilliseconds` will clear up both of these concerns in one go?

`convertToMilliseconds` speaks more to me than `convertToDate`, though it also doesn't include the full picture. What do we include them from? And then comes the more general question – can't we solve the conversation problem in a better way than copy/pasting the same function in each `from-api.js`?

Reply:
> What are the trade-offs of reusing + more complex code vs. simpler and duplicated? What should we take into account in order to make a decision?

I think that simpler and duplicated is easier to read and understand quickly - there's one file that defines a valid API response for that endpoint. You can read that file and know what data to expect.

Conversely, to read the reused code you'll need to read two files and then look at how they're put together; you'll also need to account for any conditionals to understand what is being validated against. 

On the plus side, the code that is re-used will only need to be changed once should the structure change whereas on the duplicated version it'll need to be done twice.

So, when making the decision I'd look at which is going to provide a more long term benefit - which method will ultimately be easier for future developers to work with. I don't see the API changing often (though @mattwondra will have a better long term view on this) so wouldn't be concerned with having to double up on changes too much, but it's likely other developers will need to see what the API is returning so it'd be useful to them if they could have an at-a-glance example of the format they'll be working with.

Reply:
> `convertToMilliseconds` speaks more to me than `convertToDate`, though it also doesn't include the full picture. What do we include them from? And then comes the more general question – can't we solve the conversation problem in a better way than copy/pasting the same function in each `from-api.js`?

Could we mutate the data in the selector then, before we return it? Keep the data as received in the state and then multiply it by 1000 when we call `get-concierge-available-times.js`? I know it's a no-no to do that in a reducer, but can't find anything to suggest that's bad in a selector so far.

Reply:
> I think that simpler and duplicated is easier to read and understand quickly - there's one file that defines a valid API response for that endpoint. You can read that file and know what data to expect.

That's a good point. If we reuse the schema, in order to get the full picture one would need to consult two schema files and combine them. Unless we reverse the source of truth and the appointment detail pulls from the combined `initial`.

> On the plus side, the code that is re-used will only need to be changed once should the structure change whereas on the duplicated version it'll need to be done twice.

Not the extra work worries me, but the chance of changing only one place and either forgetting about the other or just not knowing that the other place exists.

> but it's likely other developers will need to see what the API is returning so it'd be useful to them if they could have an at-a-glance example of the format they'll be working with

What do you think of having the schema in the same file and not in a separate one? It would make it really at-a-glance.

Reply:
> Could we transform the data in the selector then, before we return it? Keep the data as received in the state and then multiply it by 1000 when we call `get-concierge-available-times.js`?
Of course, everything is possible, this is just code :) What are the advantages/disadvantages of transforming the data in the selector? Do we have other options than doing it [directly in the transformation](https://github.com/Automattic/wp-calypso/blob/35607135f6f31eafca87014a4baa74e4cbf7d054/client/state/data-layer/wpcom/concierge/schedules/appointments/detail/from-api.js#L10), a shared function called by the transformation (`convertToDate`), and in the selector?

Reply:
> What do you think of having the schema in the same file and not in a separate one? It would make it really at-a-glance.

I hadn't thought of that, but yes, that makes total sense. I'll make the changes accordingly. Thanks!

Reply:
> Of course, everything is possible, this is just code :) What are the advantages/disadvantages of transforming the data in the selector? Do we have other options than doing it directly in the transformation, a shared function called by the transformation (convertToDate), and in the selector?

I've updated the PR details with a full rationale, but to summarise: we could do it in the reducer as well. I haven't, though, because the duplicated code is going to be removed in one of the PRs that follow this one so the duplication won't be an issue , it will keep the reducer nice and simple and it's consistent with how `appointment/details` works.

Doing the transformation in the selector will duplicate the work - every time the data is requested it will need to be transformed so it makes sense to do it before this.

Reply:
I wrote [my last comment](https://github.com/Automattic/wp-calypso/pull/25915#discussion_r203230965) before I saw this thread (it was hidden in GitHub because this line of code had changed).

The last thing I'd want to consider on schemas is as I noted — instead of saying "Hey, the `/initial` endpoint has the canonical example of an appointment schema... but it's tucked inside a property oddly labeled `next_appointment`...", what's the feasibility of creating some canonical schema representations for various entities in a shared context? Then every endpoint can re-use from the same shared library, rather than one endpoint being the source of truth for that entity (and needing to remember "which endpoint has the canonical `appointment` schema again?")

Reply:
If we have a central schema location we'd still have the issue with having to concatenate and maybe even modify schemas on-the-fly. I still feel like that's unnecessary. We're not duplicating business logic and I'm inclined to think that it'd add more administration when it comes to maintaining and modifying code going forward.

I take @nb's point of not wanting to miss updating one schema when the structure changes, as well as having one source of truth being a benefit. Equally, it makes sense to have a schema per endpoint as it makes logical sense to a developer, as @mattwondra says. Personally I'd still favour the `schema.json` file per endpoint. I'll wait to see if there's any additional input from @nb before I make any further changes, though :)

Reply:
I've gone ahead and reverted the schema changes and I've renamed to transformation method so it's more descriptive.

Will rebase subsequent branches accordingly.

Reply:
I probably still prefer a central schema, since in the backend they come from the same place (`Concierge_Chats_Endpoint_Response_Appointment_Detail_For_Customer`), but your argument makes sense, I don't feel so strongly about that.

We have hit a few higher-level questions in this PR that are a bit out of scope, mostly around _models_: where to keep both the schema and the transformations with shared data types, in this case the appointment one. For now, I am perfectly fine with not trying to answer those questions, I feel we did enough exploration that might help solving the problems in the future.

Reply:
Good point @nb. I gave another look through and re-marking this one as Ready to Merge.